src/utils.js

```diff
@@ -10,6 +10,34 @@ const path = require("path");
  * @typedef {() => Promise<T>} Task
  */
 
+/**
+ * @param {string} filename file path without query params (e.g. `path/img.png`)
+ * @param {string} ext new file extension without `.` (e.g. `webp`)
+ * @returns {string} new filename `path/img.png` -> `path/img.webp`
+ */
+function replaceFileExtension(filename, ext) {
+  let dotIndex = -1;
+
+  for (let i = filename.length - 1; i > -1; i--) {
+    const char = filename[i];
+
+    if (char === ".") {
+      dotIndex = i;
+      break;
+    }
+
+    if (char === "/" || char === "\\") {
+      break;
+    }
+  }
+
+  if (dotIndex === -1) {
+    return filename;
+  }
+
+  return `${filename.slice(0, dotIndex)}.${ext}`;
+}
```

Comment:
Why not regexp? I think it will be faster, because you have loop here. Can you check perf?